src/main/java/de/tum/in/www1/artemis/repository/hestia/CoverageReportRepository.java

```diff
@@ -31,31 +38,56 @@ public interface CoverageReportRepository extends ArtemisJpaRepository<CoverageR
-            ORDER BY s.submissionDate DESC
-            """)
-    // TODO: rewrite this query, pageable does not work well with left join fetch, it needs to transfer all results and only page in java
-    List<CoverageReport> getLatestCoverageReportsForLegalSubmissionsForProgrammingExerciseWithEagerFileReportsAndEntries(@Param("programmingExerciseId") Long programmingExerciseId,
-            Pageable pageable);
+    @EntityGraph(type = LOAD, attributePaths = "submission")
+    List<CoverageReport> findCoverageReportsWithSubmissionByIdIn(List<Long> ids);
+
+    /**
+     * Retrieves the latest coverage reports with legal submissions for a specific programming exercise, with pagination support.
+     * This method avoids in-memory paging by retrieving the coverage report IDs directly from the database.
+     *
+     * @param programmingExerciseId the ID of the programming exercise to retrieve the coverage reports for
+     * @param pageable              the pagination information
+     * @return a list of {@code CoverageReport} with legal submissions, or an empty list if no reports are found
+     */
+    default List<CoverageReport> getLatestCoverageReportsWithLegalSubmissionsForProgrammingExercise(Long programmingExerciseId, Pageable pageable) {
+        List<Long> ids = findCoverageReportsByProgrammingExerciseId(programmingExerciseId, pageable).stream().map(CoverageReportAndSubmissionDate::coverageReport)
+                .map(DomainObject::getId).toList();
+        if (ids.isEmpty()) {
+            return List.of();
+        }
+        return findCoverageReportsWithSubmissionByIdIn(ids);
+    }
+
+    @EntityGraph(type = LOAD, attributePaths = { "submission", "fileReports", "fileReports.testwiseCoverageEntries" })
+    List<CoverageReport> findDistinctCoverageReportsWithEagerRelationshipsByIdIn(List<Long> ids);
+
+    /**
+     * Retrieves the latest coverage reports with legal submissions for a specific programming exercise, including eager loading of file reports and entries, with pagination
+     * support.
+     * This method avoids in-memory paging by retrieving the coverage report IDs directly from the database.
+     *
+     * @param programmingExerciseId the ID of the programming exercise to retrieve the coverage reports for
+     * @param pageable              the pagination information
+     * @return a list of distinct {@code CoverageReport} with eager relationships, or an empty list if no reports are found
+     */
+    default List<CoverageReport> getLatestCoverageReportsForLegalSubmissionsForProgrammingExerciseWithEagerFileReportsAndEntries(Long programmingExerciseId, Pageable pageable) {
+        List<Long> ids = findCoverageReportsByProgrammingExerciseId(programmingExerciseId, pageable).stream().map(CoverageReportAndSubmissionDate::coverageReport)
+                .map(DomainObject::getId).toList();
+        if (ids.isEmpty()) {
+            return List.of();
```

Comment:
I suggest you unify your usage of `List.of()` vs. `Collections.emptyList()` vs. anything else you might use.